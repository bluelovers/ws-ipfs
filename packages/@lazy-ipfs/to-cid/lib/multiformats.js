"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toMultiformatsCID = void 0;
const tslib_1 = require("tslib");
const err_code_1 = (0, tslib_1.__importDefault)(require("err-code"));
const multiformats_1 = require("multiformats");
const js_cids_1 = require("@lazy-ipfs/detect-cid-lib/lib/js-cids");
const js_multiformats_1 = require("@lazy-ipfs/detect-cid-lib/lib/js-multiformats");
const util_1 = require("@lazy-ipfs/detect-cid-lib/lib/util");
const parsePath_1 = require("@lazy-ipfs/parse-ipfs-path/lib/parsePath");
const _handleLibCID_1 = require("./_handleLibCID");
function toMultiformatsCID(cidInput, libCID) {
    libCID = (0, _handleLibCID_1._handleLibCID)(libCID, multiformats_1.CID);
    if (typeof cidInput === 'string') {
        return libCID.parse((0, parsePath_1.parsePath)(cidInput, {
            unsafeReturn: true,
            noThrow: true,
        }).hash);
    }
    else if ((0, util_1._isArrayLike)(cidInput)) {
        return libCID.decode(cidInput);
    }
    let cid = libCID.asCID(cidInput);
    if ((typeof cid === 'undefined' || cid === null)) {
        if ((0, js_cids_1.isRawJsCIDLike)(cidInput)) {
            cid = libCID.asCID((0, js_cids_1.toRawJsCIDFake)(cidInput));
        }
        else if ((0, js_multiformats_1.isRawMultiformatsCIDLike)(cidInput)) {
            cid = libCID.asCID((0, js_multiformats_1.toRawMultiformatsCIDFake)(cidInput));
        }
        else if ((0, parsePath_1.isParsePathResultLoose)(cidInput)) {
            cid = toMultiformatsCID(cidInput.hash);
        }
    }
    if (!cid) {
        throw (0, err_code_1.default)(new TypeError(`Invalid type for convert to MultiformatsCID`), {
            input: cidInput,
        });
    }
    return cid;
}
exports.toMultiformatsCID = toMultiformatsCID;
exports.default = toMultiformatsCID;
//# sourceMappingURL=multiformats.js.map