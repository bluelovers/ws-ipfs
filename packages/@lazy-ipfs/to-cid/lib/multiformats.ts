import err_code from 'err-code';
import { CID as MultiformatsCID } from 'multiformats'
import { isRawJsCIDLike, SymbolJsCID, toRawJsCIDFake } from '@lazy-ipfs/detect-cid-lib/lib/js-cids';
import { ICIDValueInput, ICIDValueOrRaw } from '@lazy-ipfs/detect-cid-lib/lib/types';
import { IRawMultiformatsCIDFake, isRawMultiformatsCIDLike, toRawMultiformatsCIDFake } from '@lazy-ipfs/detect-cid-lib/lib/js-multiformats';
import { _isArrayLike } from '@lazy-ipfs/detect-cid-lib/lib/util';
import { parsePath } from '@lazy-ipfs/parse-ipfs-path/lib/parsePath';
import { _handleLibCID } from './_handleLibCID';
import { EnumTypeofCID } from '@lazy-ipfs/detect-cid-lib';

export function toMultiformatsCID<T extends ICIDValueInput, C extends MultiformatsCID = MultiformatsCID>(cidInput: T, libCID?: Pick<typeof MultiformatsCID, 'parse' | 'decode' | 'asCID'> | EnumTypeofCID): C
{
	libCID = _handleLibCID(libCID, MultiformatsCID);

	if (typeof cidInput === 'string')
	{
		return libCID.parse(parsePath(cidInput, {
			unsafeReturn: true,
			noThrow: true,
		}).hash) as any
	}
	else if (_isArrayLike(cidInput))
	{
		return libCID.decode(cidInput) as any
	}

	let cid = libCID.asCID(cidInput);

	if ((typeof cid === 'undefined' || cid === null))
	{
		if (isRawJsCIDLike(cidInput))
		{
			cid = libCID.asCID(toRawJsCIDFake(cidInput));
		}
		else if (isRawMultiformatsCIDLike(cidInput))
		{
			cid = libCID.asCID(toRawMultiformatsCIDFake(cidInput))
		}
	}

	if (!cid)
	{
		throw err_code(new TypeError(`Invalid type for convert to MultiformatsCID`), {
			input: cidInput,
		})
	}

	return cid as any
}

export default toMultiformatsCID
