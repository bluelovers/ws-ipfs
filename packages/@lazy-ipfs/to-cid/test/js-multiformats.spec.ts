import toCID, { toRawCID } from '../index';
import JsCID from 'cids';
import { CID as MultiformatsCID } from 'multiformats';

const libCID = MultiformatsCID;

test(`toCID:string`, () =>
{
	let cid = 'Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a';

	let actual = toCID(cid, libCID);
	let expected = cid;

	expect(actual.toV0().toString()).toStrictEqual(cid);

	expect(actual).toMatchSnapshot({
		multihash: {
			bytes: expect.any(Uint8Array),
		},
	});

	expect(actual.toV0().toString()).toMatchSnapshot();
	expect(actual.toV1().toString()).toMatchSnapshot();

});

test(`toCID:CID`, () =>
{
	let cid = 'Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a';

	let expected = toCID(cid, libCID);
	let actual = toCID(expected, libCID);

	expect(actual.toV0().toString()).toStrictEqual(cid);
	expect(actual).toMatchSnapshot(expected as any);

	expect(actual.toV0().toString()).toMatchSnapshot();
	expect(actual.toV1().toString()).toMatchSnapshot();

});

test(`toCID:raw`, () =>
{
	let cid = 'Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a';

	let expected = toCID(cid, libCID);
	let actual = toCID({
		"version": 0,
		"codec": "dag-pb",
		"multibaseName": "base58btc",
		multihash: Uint8Array.from([
			18,  32, 183, 254,   8,  30, 244,  17,
			96, 165, 123,  89,  19,  86,  24,  96,
			118, 229, 238, 199, 116,   2,  56,  83,
			37, 188,  26,   8,  22, 181, 187, 118,
			74, 219
		]),
	}, libCID);

	expect(actual.toV0().toString()).toStrictEqual(cid);
	expect(actual).toMatchSnapshot(expected as any);

	expect(actual.toV0().toString()).toMatchSnapshot();
	expect(actual.toV1().toString()).toMatchSnapshot();

});

test(`toCID:raw without multibaseName`, () =>
{
	let cid = 'Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a';

	let expected = toCID(cid, libCID);
	let actual = toCID({
		"version": 0,
		"codec": "dag-pb",
		multihash: Uint8Array.from([
			18,  32, 183, 254,   8,  30, 244,  17,
			96, 165, 123,  89,  19,  86,  24,  96,
			118, 229, 238, 199, 116,   2,  56,  83,
			37, 188,  26,   8,  22, 181, 187, 118,
			74, 219
		]),
	}, libCID);

	expect(actual.toV0().toString()).toStrictEqual(cid);
	expect(actual).toMatchSnapshot(expected as any);

	expect(actual.toV0().toString()).toMatchSnapshot();
	expect(actual.toV1().toString()).toMatchSnapshot();

});
