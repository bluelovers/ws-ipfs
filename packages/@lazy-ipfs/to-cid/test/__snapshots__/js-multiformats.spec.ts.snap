// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toCID:CID 1`] = `
{
  "bytes": {
    "0": 18,
    "1": 32,
    "10": 123,
    "11": 89,
    "12": 19,
    "13": 86,
    "14": 24,
    "15": 96,
    "16": 118,
    "17": 229,
    "18": 238,
    "19": 199,
    "2": 183,
    "20": 116,
    "21": 2,
    "22": 56,
    "23": 83,
    "24": 37,
    "25": 188,
    "26": 26,
    "27": 8,
    "28": 22,
    "29": 181,
    "3": 254,
    "30": 187,
    "31": 118,
    "32": 74,
    "33": 219,
    "4": 8,
    "5": 30,
    "6": 244,
    "7": 17,
    "8": 96,
    "9": 165,
  },
  "code": 112,
  "multihash": {
    "bytes": {
      "0": 18,
      "1": 32,
      "10": 123,
      "11": 89,
      "12": 19,
      "13": 86,
      "14": 24,
      "15": 96,
      "16": 118,
      "17": 229,
      "18": 238,
      "19": 199,
      "2": 183,
      "20": 116,
      "21": 2,
      "22": 56,
      "23": 83,
      "24": 37,
      "25": 188,
      "26": 26,
      "27": 8,
      "28": 22,
      "29": 181,
      "3": 254,
      "30": 187,
      "31": 118,
      "32": 74,
      "33": 219,
      "4": 8,
      "5": 30,
      "6": 244,
      "7": 17,
      "8": 96,
      "9": 165,
    },
    "code": 18,
    "digest": {
      "0": 183,
      "1": 254,
      "10": 19,
      "11": 86,
      "12": 24,
      "13": 96,
      "14": 118,
      "15": 229,
      "16": 238,
      "17": 199,
      "18": 116,
      "19": 2,
      "2": 8,
      "20": 56,
      "21": 83,
      "22": 37,
      "23": 188,
      "24": 26,
      "25": 8,
      "26": 22,
      "27": 181,
      "28": 187,
      "29": 118,
      "3": 30,
      "30": 74,
      "31": 219,
      "4": 244,
      "5": 17,
      "6": 96,
      "7": 165,
      "8": 123,
      "9": 89,
    },
    "size": 32,
  },
  "version": 0,
}
`;

exports[`toCID:CID 2`] = `"Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a"`;

exports[`toCID:CID 3`] = `"bafybeifx7yeb55armcsxwwitkymga5xf53dxiarykms3ygqic223w5sk3m"`;

exports[`toCID:raw 1`] = `
{
  "bytes": {
    "0": 18,
    "1": 32,
    "10": 123,
    "11": 89,
    "12": 19,
    "13": 86,
    "14": 24,
    "15": 96,
    "16": 118,
    "17": 229,
    "18": 238,
    "19": 199,
    "2": 183,
    "20": 116,
    "21": 2,
    "22": 56,
    "23": 83,
    "24": 37,
    "25": 188,
    "26": 26,
    "27": 8,
    "28": 22,
    "29": 181,
    "3": 254,
    "30": 187,
    "31": 118,
    "32": 74,
    "33": 219,
    "4": 8,
    "5": 30,
    "6": 244,
    "7": 17,
    "8": 96,
    "9": 165,
  },
  "code": 112,
  "multihash": {
    "bytes": {
      "0": 18,
      "1": 32,
      "10": 123,
      "11": 89,
      "12": 19,
      "13": 86,
      "14": 24,
      "15": 96,
      "16": 118,
      "17": 229,
      "18": 238,
      "19": 199,
      "2": 183,
      "20": 116,
      "21": 2,
      "22": 56,
      "23": 83,
      "24": 37,
      "25": 188,
      "26": 26,
      "27": 8,
      "28": 22,
      "29": 181,
      "3": 254,
      "30": 187,
      "31": 118,
      "32": 74,
      "33": 219,
      "4": 8,
      "5": 30,
      "6": 244,
      "7": 17,
      "8": 96,
      "9": 165,
    },
    "code": 18,
    "digest": {
      "0": 183,
      "1": 254,
      "10": 19,
      "11": 86,
      "12": 24,
      "13": 96,
      "14": 118,
      "15": 229,
      "16": 238,
      "17": 199,
      "18": 116,
      "19": 2,
      "2": 8,
      "20": 56,
      "21": 83,
      "22": 37,
      "23": 188,
      "24": 26,
      "25": 8,
      "26": 22,
      "27": 181,
      "28": 187,
      "29": 118,
      "3": 30,
      "30": 74,
      "31": 219,
      "4": 244,
      "5": 17,
      "6": 96,
      "7": 165,
      "8": 123,
      "9": 89,
    },
    "size": 32,
  },
  "version": 0,
}
`;

exports[`toCID:raw 2`] = `"Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a"`;

exports[`toCID:raw 3`] = `"bafybeifx7yeb55armcsxwwitkymga5xf53dxiarykms3ygqic223w5sk3m"`;

exports[`toCID:raw without multibaseName 1`] = `
{
  "bytes": {
    "0": 18,
    "1": 32,
    "10": 123,
    "11": 89,
    "12": 19,
    "13": 86,
    "14": 24,
    "15": 96,
    "16": 118,
    "17": 229,
    "18": 238,
    "19": 199,
    "2": 183,
    "20": 116,
    "21": 2,
    "22": 56,
    "23": 83,
    "24": 37,
    "25": 188,
    "26": 26,
    "27": 8,
    "28": 22,
    "29": 181,
    "3": 254,
    "30": 187,
    "31": 118,
    "32": 74,
    "33": 219,
    "4": 8,
    "5": 30,
    "6": 244,
    "7": 17,
    "8": 96,
    "9": 165,
  },
  "code": 112,
  "multihash": {
    "bytes": {
      "0": 18,
      "1": 32,
      "10": 123,
      "11": 89,
      "12": 19,
      "13": 86,
      "14": 24,
      "15": 96,
      "16": 118,
      "17": 229,
      "18": 238,
      "19": 199,
      "2": 183,
      "20": 116,
      "21": 2,
      "22": 56,
      "23": 83,
      "24": 37,
      "25": 188,
      "26": 26,
      "27": 8,
      "28": 22,
      "29": 181,
      "3": 254,
      "30": 187,
      "31": 118,
      "32": 74,
      "33": 219,
      "4": 8,
      "5": 30,
      "6": 244,
      "7": 17,
      "8": 96,
      "9": 165,
    },
    "code": 18,
    "digest": {
      "0": 183,
      "1": 254,
      "10": 19,
      "11": 86,
      "12": 24,
      "13": 96,
      "14": 118,
      "15": 229,
      "16": 238,
      "17": 199,
      "18": 116,
      "19": 2,
      "2": 8,
      "20": 56,
      "21": 83,
      "22": 37,
      "23": 188,
      "24": 26,
      "25": 8,
      "26": 22,
      "27": 181,
      "28": 187,
      "29": 118,
      "3": 30,
      "30": 74,
      "31": 219,
      "4": 244,
      "5": 17,
      "6": 96,
      "7": 165,
      "8": 123,
      "9": 89,
    },
    "size": 32,
  },
  "version": 0,
}
`;

exports[`toCID:raw without multibaseName 2`] = `"Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a"`;

exports[`toCID:raw without multibaseName 3`] = `"bafybeifx7yeb55armcsxwwitkymga5xf53dxiarykms3ygqic223w5sk3m"`;

exports[`toCID:string 1`] = `
{
  "bytes": Uint8Array [
    18,
    32,
    183,
    254,
    8,
    30,
    244,
    17,
    96,
    165,
    123,
    89,
    19,
    86,
    24,
    96,
    118,
    229,
    238,
    199,
    116,
    2,
    56,
    83,
    37,
    188,
    26,
    8,
    22,
    181,
    187,
    118,
    74,
    219,
  ],
  "code": 112,
  "multihash": {
    "bytes": Any<Uint8Array>,
    "code": 18,
    "digest": Uint8Array [
      183,
      254,
      8,
      30,
      244,
      17,
      96,
      165,
      123,
      89,
      19,
      86,
      24,
      96,
      118,
      229,
      238,
      199,
      116,
      2,
      56,
      83,
      37,
      188,
      26,
      8,
      22,
      181,
      187,
      118,
      74,
      219,
    ],
    "size": 32,
  },
  "version": 0,
}
`;

exports[`toCID:string 2`] = `"Qmaisz6NMhDB51cCvNWa1GMS7LU1pAxdF4Ld6Ft9kZEP2a"`;

exports[`toCID:string 3`] = `"bafybeifx7yeb55armcsxwwitkymga5xf53dxiarykms3ygqic223w5sk3m"`;
